core/src/main/scala/org/apache/gearpump/cluster/master/AppManager.scala

```diff
@@ -0,0 +1,256 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.cluster.master
+
+import java.util.concurrent.{TimeUnit, TimeoutException}
+
+import akka.actor._
+import akka.cluster.Cluster
+import akka.pattern.ask
+import org.apache.gearpump.cluster.AppMasterToMaster._
+import org.apache.gearpump.cluster.AppMasterToWorker._
+import org.apache.gearpump.cluster.ClientToMaster._
+import InMemoryKVService._
+import MasterHAService._
+import org.apache.gearpump.cluster.MasterToAppMaster._
+import org.apache.gearpump.cluster.MasterToClient.{ReplayApplicationResult, ResolveAppIdResult, ShutdownApplicationResult, SubmitApplicationResult}
+import org.apache.gearpump.cluster.WorkerToAppMaster._
+import org.apache.gearpump.cluster._
+import org.apache.gearpump.cluster.scheduler.{Resource, ResourceAllocation, ResourceRequest}
+import org.apache.gearpump.transport.HostPort
+import org.apache.gearpump.util.ActorSystemBooter._
+import org.apache.gearpump.util.Constants._
+import org.apache.gearpump.util._
+import org.slf4j.Logger
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.concurrent.duration.Duration
+import scala.util.{Failure, Success}
+
+/**
+ * AppManager is dedicated part of Master to manager applicaitons
+ */
+private[cluster] class AppManager(masterHA : ActorRef, kvService: ActorRef, launcher: AppMasterLauncherFactory) extends Actor with Stash with TimeOutScheduler{
+  private val LOG: Logger = LogUtil.getLogger(getClass)
+
+  private val executorId : Int = APPMASTER_DEFAULT_EXECUTOR_ID
+  private val systemconfig = context.system.settings.config
+
+  implicit val timeout = FUTURE_TIMEOUT
+  implicit val executionContext = context.dispatcher
+
+  //next available appId
+  private var appId: Int = 0
+
+  //from appid to appMaster data
+  private var appMasterRegistry = Map.empty[Int, (ActorRef, AppMasterRuntimeInfo)]
+
+  def receive: Receive = null
+
+  (masterHA ? GetMasterState).asInstanceOf[Future[GetMasterStateResult]].map(self ! _)
+  context.become(waitForMasterState)
+
+  def waitForMasterState: Receive = {
+    case MasterState(state) =>
+      val set = state.map(_.appId)
+      if (set.isEmpty) {
+        this.appId = 0
+      } else {
+        this.appId = state.map(_.appId).max + 1
```

Comment:
Maybe it is better to store max appId separately. Otherwise, we may generate duplicate application ID if the max appId used but the app ended and appmanager restarted.